### ISSUE TITLE
[Joins] Combining 2 indices for filtering and sorting

### ISSUE THREAD

**Issue by joostmeijles** [typesense/typesense#229]
I am looking for a efficient way to filter and sort on personalised pricing for a product catalog.
The product catalog has ~50K products. Each product has a personalised price per customer. There are around ~40K customers.

My (first) idea is to have one index for all products and one price index per customer.
These 2 indices I want to use to filter and sort on pricing per customer.

Would this be possible with Typesense?

NB. I know about the multiple search functionality of Typesense, but that does not seem to fulfil my needs.

**Comment by jasonbosco** [comment]
Hi @joostmeijles, Typesense doesn't support doing a SQL-esque "join" across two indices. So you'd have to essentially pre-join the data. 

Here are two ways to do it:

1. Create one document per product per customer (with the customized price embedded) and put all these documents in the same index. So something like this:

```json
[
  { "product_id": 1, "customer_id": 1, "product_price": 1000, "product_name": "A" },
  { "product_id": 2, "customer_id": 1, "product_price": 2000, "product_name": "B" },
  { "product_id": 1, "customer_id": 2, "product_price": 1003, "product_name": "A" },
  { "product_id": 2, "customer_id": 2, "product_price": 2003, "product_name": "B" },
]
```

You can then generate a [scoped API key](https://typesense.org/docs/0.19.0/api/#generate-scoped-search-key) with the `customer_id` embedded in it, to only allow a customer to see their own product records. 

Now this would create an index that has `50K * 40K = 2B` records. I haven't tested Typesense with this size of a dataset in a single collection, but I'd be curious to see how it performs! From an index size perspective, you'll see very efficient RAM usage because the documents have repeated field values in them and so the index will "compress" very well. 

2. Create one collection per customer, with their personalized prices embedded in each product record in each collection. 

The smaller record count per customer in (2) will result in a smaller index size per customer, but we're essentially duplicating product data, so overall RAM usage won't be as efficient.

**Comment by joostmeijles** [comment]
Hi @jasonbosco, there are ~200 attributes per product would that still be memory efficient when using one collection?

As the prices are updated regularly, e.g. once per day, do you have an idea about the ingest (and full-index / incremental-index update) performance?

**Comment by jasonbosco** [comment]
@joostmeijles Are all 200 attributes required to be searchable by your users? You only want to add fields that are user-searchable/filterable to the collection schema and only these will be indexed (in memory). The rest of the fields will just be stored on disk and returned whenever a record is a search hit, and so won't consume memory. In any case, Option (1) is generally more memory efficient than Option (2).

Now, updating 2B records is something I haven't tried before (due to not having a real-world dataset of that scale). So you'll have to tell me how things go, given the particular shape and size of your data. One thing I would recommend is using the `import` endpoint and batch updating records, instead of updating them one by one, to avoid the HTTP request overhead. 

On a side note, it's interesting how a seemingly benign size of a dataset with 50K products can result in 2B records with personalized prices. So we're thinking of exploring support for simple joins between collections, to support use cases like yours better...

**Comment by joostmeijles** [comment]
>On a side note, it's interesting how a seemingly benign size of a dataset with 50K products can result in 2B records with personalized prices. So we're thinking of exploring support for simple joins between collections, to support use cases like yours better...

That would be great! 

I think this article https://medium.com/vespa/efficient-personal-search-at-large-scale-a1e8250ec945 about personalized search (using Vespa) describes the challenge pretty nice.

**Comment by nickjackson** [comment]
We also have this exact use case for one client (custom user pricing per product) but we have just hit another scenario where join's could be useful.

The model is similar to Github where we have concepts similar to “users” and “repos”. We want to quickly search through repos and display the results which have a link through to the repo page. However, the problem is that we have usernames in the links, and those usernames can change. If a user wants to change their username, we have to re-index the documents to update the links.

If we had joins, we could have two collections, “users”, and “repos”. We would run our searches against the "repos" collection, and request Typesense resolve the `user_id` into a full object or certain fields.

**Comment by iSuslov** [comment]
I think this will be a killer feature. It will be so convenient not to prepare data by nesting before uploading it to typesense.

**Comment by ViktorQvarfordt** [comment]
This would solve our problem:

We struggle with allowing users to search across private content. We have both these situations:
* A user can have thousands of documents they can privately access.
* A document can have thousands of users that can privately access it.

Both approaches I know of suffer from limitations:
* Put a field `accessible_to_user_ids` of type `string[]` on each document and user a filter `filter_by: accessible_to_user_ids:=${userId}`. The problem is that the array `accessible_to_user_ids` can be extremely long causing huge documents.
* Use a filter `filter_by: id:=${documentIdsAvailableToUser}`. The problem is that the filter can become extremely large due to the size of `documentIdsAvailableToUser`.

Slack thread: https://typesense-community.slack.com/archives/C01P749MET0/p1672839899898999

**Comment by iSuslov** [comment]
@victorbosh isn't it a [Scoped Search API Key](https://typesense.org/docs/0.23.1/api/api-keys.html#generate-scoped-search-key) problem?
https://typesense.org/docs/guide/data-access-control.html

**Comment by shaunbyrnes** [comment]
Just checking how this feature is progessing, it will help us with **a lot of things** which are currently much more convoluted without the ability to connect/join collections. For example when searching Typesense colletions which have several ID records to other collections with data which the user needs to see (like brands/categories), we would have to make those per record as async calls on client than just getting them from other Typesense collections.

**Comment by jasonbosco** [comment]
We're about 80% done with this feature, but we had to pause work on it to focus on other time-sensitive priorities. We plan to resume work on this in a month or so. 

In the meantime, @shaunbyrnes for your specific use-case, I'd recommend storing the actual brand names and category names in each record, instead of the ID. Repeated field values across records compress well when indexed in Typesense.

**Comment by shaunbyrnes** [comment]
@jasonbosco That's good news you guys are making progress on it. Understood on the other other priorities. 

We had considered that as an alternative, but as the source data in Firebase only has the ID/foreign key, it isn't possible for the name to come across - unless we don't use the marketplace extension and its auto created ext cloud function, but instead create a custom cloud function to watch in Firebase and write the changes. It also means that any brand/category name updates for example, which happen in their respective source collections, wouldn't then be reflected in Typesense either.

Hence the join type ability would solve this completely. Just like the requirements we have for personalised data where the user's stats around shared global content isn't practical to store all in one place - they would need to be seperate entities and joined.

**Comment by DigitalVanilla** [comment]
+1, join collections would be a very useful feature (ditto)

**Comment by shaunbyrnes** [comment]
@jasonbosco @kishorenc how are we progressing with this?

**Comment by brentspg** [comment]
Agreed. This would be a very helpful feature add for our platform.

**Comment by sebmarz** [comment]
Yes, this will be for sure helpful for us with the same requirements as @joostmeijles

**Comment by kishorenc** [comment]
We're actively working on this, you can expect a test build in 2-3 weeks.

**Comment by shaunbyrnes** [comment]
Sounds good @kishorenc thanks for the update!

**Comment by muneebshahid** [comment]
Hey @kishorenc thank you for the effort you guys are putting in. Has there been any progress on this?

**Comment by kishorenc** [comment]
Here's a Docker build that introduces support for joins:

```
typesense/typesense:0.26.0.rc8
```

Instructions: https://gist.github.com/kishorenc/037314619c3104970ae703b25aff2f51

We support joins at the moment, but will be updating the instructions to add more advance use cases soon.

Please give it a shot and let us know what you think. Happy to help you all address your use cases.

**Comment by shaunbyrnes** [comment]
Awesome, thanks for the update Kishore!1) Can we apply this build to Typesense Cloud (where all our data is), to test?2) In the example there was just one referenced collection and field shown. Are more than 1 supported, by using additional $CollectionName() syntax?Cheers! On 21 Jul 2023, at 11:19 pm, Kishore Nallan ***@***.***> wrote:﻿
Here's a Docker build that introduces support for joins:
typesense/typesense:0.26.0.rc8

Instructions: https://gist.github.com/kishorenc/037314619c3104970ae703b25aff2f51
We support joins at the moment, but will be updating the instructions to add more advance use cases soon.
Please give it a shot and let us know what you think. Happy to help you all address your use cases.

**Comment by kishorenc** [comment]
@shaunbyrnes 

Happy to upgrade your cluster on TS Cloud to this version as well, but you should probably try this on a staging cluster first. If you have one, please email contact@typesense.org with the cluster ID.

>  In the example there was just one referenced collection and field shown. Are more than 1 supported, by using additional $CollectionName() syntax?

We are finishing up some additional updates for N-way joins. Can you give an example of the type of query you are looking to do? I can then suggest how to achieve that.

**Comment by shaunbyrnes** [comment]
@kishorenc makes sense, sure.
So example situation is 2 collections `food` and `meals` which have different but similar schemas and we would like the results from both of them to be combined in certain situations, such as a results list of "my food and meals" basically combination of both collections which match the current items created by the user logged in.

Furthermore, user's can favorite food or meals, which are currently stored in collections (in Firestore) `favoriteMeals` and `favoriteFood` referencing these respective collections would also be part of the plan, as we have a "favorites" screen which shows a combined list of both foods _and_ meals which the logged in user has added to their favorites. 

Currently to accomplish these screens requires numerous queries sequentially/concurrently to gather all the data then collating it on client side and much more complicated pagination logic etc. To further complicate matters, the use case allows for not only text searching but other filters.

A final missing requirement is like the first use case example given on your link; food and/or meals have stats associated with them such as frequency or last used etc. These need to live outside the food/meals collections related to each user. We need to also reference the userFoodMealStats collection to integrate this with our querying! 

Our aim would be to reference these collections in a single query as needed and let TS link up the data, then return after all filtering/sort etc. is done server side.

**Comment by shaunbyrnes** [comment]
Did the comment above make sense @kishorenc ?

**Comment by kishorenc** [comment]
@shaunbyrnes 

I suspect that what you've described will require both union and join. We only support joins at the moment where records sharing a common ID but in two different collections can be joined as one.

**Comment by shaunbyrnes** [comment]
Yes got it, but wouldnt we just join via `userId` between all collections? Both food and meals will have userId, as will the favorites collections and the user stats. They are all common around the user. 

You mentioned union, you have plans for this feature too? How would it differ than the join functionality just added?

**Comment by joostmeijles** [comment]
Very nice progress! Are there any limitations on the number of collections (within a node)?

**Comment by jasonbosco** [comment]
> Very nice progress! 

Thank you!

> Are there any limitations on the number of collections (within a node)?

No - I've seen users with hundreds of thousands of collections per node.

**Comment by danthegoodman1** [comment]
This would be amazing. This feature was the primary reason we tried meilisearch first (but ultimately it couldn't handle our workload)

**Comment by jasonbosco** [comment]
> This feature was the primary reason we tried meilisearch first

Out of curiosity, may I know which feature you’re referring to? If it’s the join feature described in this thread, I don’t think Meilisearch has such a feature IIRC?

**Comment by danthegoodman1** [comment]
They do, with meilisearch when you query multiple indexes they rank together, where as with typesense they rank separately per index https://www.meilisearch.com/docs/reference/api/multi_search

**Comment by jasonbosco** [comment]
> with meilisearch when you query multiple indexes they rank together

Interesting. This sounds new, but my understanding of that feature is that it's still not doing cross-index ranking, based on this description in the docs:

> Results of the search queries in the same order they were requested in

It's just that the format of the response is in a single array (still not ranked across indices). 

Anywho, that is slightly different from the feature discussed in this thread. This is more like `JOIN ... ON (...) WHERE` in SQL. Multi-search is more like UNION in SQL.

**Comment by danthegoodman1** [comment]
Well even better for typesense then!

**Comment by yoshig** [comment]
Will it be possible to use these joins for searches? For example, we want to search for contact names, where contacts will be joined to a company.

**Comment by yoshig** [comment]
Another request, there is functionality that we were trying to replace, such as `$specialties(specialty_names:!=[Foo,Bar])`. These queries seem to stall out (taking more than 16 seconds to complete).

**Comment by comichael** [comment]
I understand JOINs in Typesense using "reference" are not officially supported yet (since there is no info in the official documentation). Is that right? If so, do you have any plan to include it anytime soon?

We have about 1 million products, and 10 warehouses. New warehouses can be created at any time. In product search, we need to show the product stock for each warehouse. What would be the best way to implement this in Typesense (using the dotnet library)? Would you recommend a flat structure (Product 1 - Qty Warehouse A - Qty Warehouse B) or some solution using facets (which I have not learned too much about yet, but I will if that is the best solution).

**Comment by happy-san** [comment]
@comichael Regarding the implementation, I would suggest you create the following three collections:
* Products
* Warehouses
* WarehouseProductQuantity

`WarehouseProductQuantity` is a junction collection that will reference both a warehouse and a product. So every document in this collection will represent the quantity of a product at a particular warehouse. In this configuration, you'll be free to add/remove products/warehouses.
Search request will look like:
```json
{
  "collection": "Products",
  "q": "foo",
  "query_by": "product_name",
  "filter_by": "$WarehouseProductQuantity(id: *)",
  "include_fields": "$Warehouses(warehouse_name)"
}
```
[More more info](https://gist.github.com/kishorenc/037314619c3104970ae703b25aff2f51)

**Comment by comichael** [comment]
Thx @happy-san!

> "WarehouseProductQuantity" is a junction collection that will reference both a warehouse and a product. 

This  would be the perfect solution for us, since it reflects the existing DB structure, more or less.

However - Is it correct that to implement this, we would need to reinstall our TS server with a docker image like 0.26.0.rc8, and we must use an alternative to typesense-dotnet until they have implemented the concept of "references"?

**Comment by happy-san** [comment]
> we would need to reinstall our TS server with a docker image like 0.26.0.rc8
Yes
> use an alternative to typesense-dotnet until they have implemented the concept of "references"
I think they'll only need to add a `reference` field in the `Field` class.

**Comment by Mat-moran** [comment]
Hi, Amazing feature.

I was trying to find a solution for a pricelist per user problem. 

I am using type sense cloud. Any idea of when will be available for the cloud?

Thanks

**Comment by Mat-moran** [comment]
I am testing it, I notice that it is possible to upgrade typesense cloud to the latest version.

Currently I am testing the https://github.com/typesense/typesense-instantsearch-adapter , it is compatible with this feature?
I am getting `Error: 400 - Could not parse the filter query.` in the client when I try to configure it with a predefined filter.
```typescript
<InstantSearch
        indexName="products"
        searchClient={searchClient}
      >
        <Configure
          hitsPerPage={2}
          facetFilters={["$pricelist(pricelist_id:=1)"]}
        />
        <SearchBox />
        <SearchAndFilter />
        <Hits hitComponent={Hit} />
      </InstantSearch>
    </>
```
The same filter is working properly through the API.

thanks 😄

**Comment by jasonbosco** [comment]
@Mat-moran `facetFilters` needs to be in Algolia's filter format, and Algolia does not support this feature. So you want to use the `filters` prop in the `Configure` widget, and then you should be able to use the Typesense syntax inside. 

Could you try this:

```jsx
<Configure
  hitsPerPage={2}
  filters={"$pricelist(pricelist_id:=1)"}
/>
```

**Comment by Mat-moran** [comment]
Thanks!  it works perfectly 😄

**Comment by jasonbosco** [comment]
Released in v26

**Comment by brettwillis** [comment]
I'm having a couple of issues with this feature, maybe you can tell me what I'm doing wrong?

Consider two hypothetical collections like so

```json
{
  "name": "users",
  "fields": [
    { "name": "id", "type": "string" },
    { "name": "name", "type": "string" },
    { "name": "orders", "type": "string[]", "reference": "orders.id" }
  ]
}
```

```json
{
  "name": "orders",
  "fields": [
    { "name": "id", "type": "string" },
    { "name": "user", "type": "string", "reference": "users.id" },
  ]
}
```

## (1) Missing relations are not allowed

The error `Referenced document having 'id: ...' not found in the collection '...'.` has the following implications
1. Seems impossible to have two-way relation like above. One document cannot exist before the other.
2. Reverting to a one-way relation to work around (1), I actually want to batch-import both of the `users` and `orders` collections in parallel. But I cannot because it is unreliable to know which of the users/orders will be imported/committed first.

To explain, my application does regular large batch imports, and the source data actually has the `orders` embedded into the `user` document e.g. `{id: 'user123', name: '...', orders: [{...}, {...}]}`. So it is actually a bit of an indirection for me to break out the `orders` into a separate collection for Typesense. However I need to do this to utilise the filtering/sorting on orders which I otherwise cannot do if `orders` was just an array of nested objects in the `user`.

So now, because Typesense throws on missing relations, I cannot import both collections in parallel, but hold the entire batch in memory while I import `orders` and then `users`.

Questions:
1. Is it possible to silently fail/omit missing relations so that this whole workflow is much more flexible?
2. Feature request: Is it possible to have the same filter/sort on an array of nested objects identical to the JOIN functionality? I.e. treat array of nested objects like a pre-join (if my wording makes sense)? Filtering by nested fields in arrays is already supported to some degree, but not sorting by nested fields like you now can with joins.

## (2) Peculiar join/populate behaviour with arrays

In the example above, if I do `.../search?q=&query_by=name&include_fields=$orders(*,strategy:nest_array) as orders` then I get the following document for users

```json
{
  "id": "user123",
  "name": "..."
  "orders": [
    "order1",
    "order2",
    { "id": "order1" },
    { "id": "order2" },
  ]
}
```

I.e. the joined/populated orders are appended to the array of string IDs, rather than replacing the ID.

Also the relation is not automatically populated like the docs suggest, I need to force it with the `include_fields` param.

**Comment by brettwillis** [comment]
Also I'm getting a `Could not find a reference for doc 0` when sorting on some joined collections but it works fine for joins referring to other collections. What does this error mean? Is there a broken reference? This even happens when there are zero items/references in the actual data.

**Comment by happy-san** [comment]
@brettwillis Regarding the import issue, can you populate a list of `user_id, order_id` pair during the batch import? If yes, then you can create a junction collection `users_orders` with the following schema
```json
{
  "name": "users_orders",
  "fields": [
    { "name": "user_id", "type": "string", "reference": "users.id" },
    { "name": "order_id", "type": "string", "reference": "orders.id" }
  ]
}
```
and you can import that list after the batch import has been completed.

> Is it possible to silently fail/omit missing relations so that this whole workflow is much more flexible?
No, this is not possible. We create references only during the indexing that are used during joins. We allow reference fields to be absent with `optional: true` but that would still fail if the value can't be found in the referenced collection.

**Comment by happy-san** [comment]
> Feature request: Is it possible to have the same filter/sort on an array of nested objects identical to the JOIN functionality? I.e. treat array of nested objects like a pre-join (if my wording makes sense)? Filtering by nested fields in arrays is already supported to some degree, but not sorting by nested fields like you now can with joins.

Can you explain this with an example? Typesense allows sorting only on singular values. So when you have multiple references in a single doc that would also fail to sort by on a reference field.

**Comment by brettwillis** [comment]
> Typesense allows sorting only on singular values. So when you have multiple references in a single doc that would also fail to sort by on a reference field.

Ok does this issue apply for joins too? E.g. `/users/search?q=&query_by=name&sort_by=$orders(createdAt:desc)` such a search I would want to sort users by those have have the most recent order first. And I would want the same behaviour for for array/embedded values, but as you say Typesense does not allow sorting on embedded array values right now...

> Can you explain this with an example?

@happy-san ok I've actually got several use cases that I'm struggling to find a good structure for with Typesense, and I has excited that joins would give me some fresh potential. I would like to discuss it but perhaps this issue is not the right place (could end up a rather large discussion), should I open another issue?

**Comment by happy-san** [comment]
Yes `/users/search?q=&query_by=name&sort_by=$orders(createdAt:desc)` would fail too. #867 is for this feature.

Opening a new issue for your use case is fine or maybe you can post in our community slack channel.
